ao_experimental/quant_primitives.py

```diff
@@ -0,0 +1,43 @@
+import torch
+from torch import _dynamo
+
+def safe_int_mm(x_int8: torch.Tensor, w_int8: torch.Tensor):
+    """
+    This function wraps torch._int_mm and avoids several undesirable behaviors of the function for certain inputs while still 
+    returning correct results and being torch.compiled in a performant way.
+
+    Assumes both tensors have dimension of 2.
+    
+    Note: no error checking for torch.compiled path, if x_int8.shape = [i, j] and j<=16 then the triton kernel
+    will silently give incorrect results
+
+    Args:
+        x_int8 (Tensor, torch.int8): the first tensor to be multiplied
+        w_int8 (Tensor, torch.int8): the second tensor to be multiplied
+
+    Return:
+        out (Tensor, torch.int32): the result of the matmul with device matching that of the inputs
+    """
+
+    # torch.compile path
+    if torch._dynamo.is_compiling():
+        return torch._int_mm(x_int8, w_int8)
+    
+    # error checking for cublas path    
+    device_cpu = 'cpu' in [w_int8.device.type, x_int8.device.type]
```

Comment:
we should ensure the device is the same or error out here, otherwise this would enable people to write code which will handle devices incorrectly, and this function will magically fix it

Reply:
will do